conductr_cli/sandbox_proxy.py

```diff
@@ -0,0 +1,107 @@
+                              '  option  dontlognull\n' \
+                              '  timeout connect 5000\n' \
+                              '  timeout client  50000\n' \
+                              '  timeout server  50000\n' \
+                              '\n' \
+                              'frontend conductr-haproxy-test\n' \
+                              '  bind :65535\n' \
+                              '  mode http\n' \
+                              '  monitor-uri /test\n' \
+                              ''
+
+
+def start_proxy(proxy_bind_addr, proxy_ports):
+    validate_docker_present()
+    setup_haproxy_dirs()
+    stop_proxy()
+    start_docker_instance(proxy_bind_addr, proxy_ports)
+    start_conductr_haproxy()
+
+
+def stop_proxy():
+    # TODO: validate docker present within try block, and catch the error and return False
+    log = logging.getLogger(__name__)
+
+    running_container = get_running_haproxy()
+
+    if running_container:
+        log.info(headline('Stopping HAProxy'))
+        validate_docker_present()
+        terminal.docker_rm([DEFAULT_SANDBOX_PROXY_CONTAINER_NAME])
+        log.info('HAProxy has been successfully stopped')
+
+    return True
+
+
+def validate_docker_present():
+    # TODO: implement docker installation validation
+    pass
+
+
+def setup_haproxy_dirs():
+    os.makedirs(HAPROXY_CFG_DIR, exist_ok=True)
+
+    # Correct haproxy.cfg file must exists in order for Docker image to be working
+    haproxy_cfg = pathlib.Path(HAPROXY_CFG_PATH)
+    if not haproxy_cfg.exists():
+        haproxy_cfg.write_text(DEFAULT_HAPROXY_CFG_ENTRIES)
+
+
+def get_running_haproxy():
+    return terminal.docker_ps(ps_filter='name={}'.format(DEFAULT_SANDBOX_PROXY_CONTAINER_NAME))
+
+
+def start_docker_instance(proxy_bind_addr, proxy_ports):
+    log = logging.getLogger(__name__)
+    log.info(headline('Starting HAProxy'))
+
+    docker_image = terminal.docker_images(HAPROXY_DOCKER_IMAGE)
+    if not docker_image:
+        log.info('Pulling docker image {}'.format(HAPROXY_DOCKER_IMAGE))
+        terminal.docker_pull(HAPROXY_DOCKER_IMAGE)
+
+    all_proxy_ports = []
+    all_proxy_ports.extend(DEFAULT_PROXY_PORTS)
+    all_proxy_ports.extend(proxy_ports)
+    all_proxy_ports = sorted(all_proxy_ports)
+
+    port_args = []
+    for port in all_proxy_ports:
+        port_args.append('-p')
+        port_args.append('{}:{}:{}'.format(proxy_bind_addr.exploded, port, port))
+
+    docker_args = ['-d', '--name', DEFAULT_SANDBOX_PROXY_CONTAINER_NAME] + port_args + \
+                  ['-v', '{}:/usr/local/etc/haproxy:ro'.format(HAPROXY_CFG_DIR)]
+
+    log.info('Exposing the following ports {}'.format(all_proxy_ports))
+    terminal.docker_run(docker_args, HAPROXY_DOCKER_IMAGE, positional_args=[])
+
+
+def start_conductr_haproxy():
+    log = logging.getLogger(__name__)
+    bundle_name = 'conductr-haproxy'
+    configuration_name = 'conductr-haproxy-dev-mode'
+    log.info(headline('Starting ConductR HAProxy'))
```

Comment:
Previously, we didn't have a separate headline for `conductr-haproxy`, meaning the deployment of `conductr-haproxy` is part of the `Starting HAProxy` section. I think this makes sense and therefore we can remove this headline here. Also note that we do have a `Stopping HAProxy` section but no `Stopping ConductR HAProxy` section (because it is stopped as part of ConductR).